src/StepperISR_due.cpp

```diff
@@ -1,401 +1,691 @@
+removed!
+
+With that, I've spun this motor to about 1600 RPM.  No matter how slow the
+acceleration, it seems to stop somewhere just above that.  So the code can do
+more than my nema23 test motor :)
+
+Last "trick".  I didn't think about the fact that at 21MHz with a 16 bit
+counter, the longest the system can delay is around 3ms.  Fortunately, this was
+not lost on the original author.  the hasSteps or steps==0 case is when you
+want to pause for a time (specified in the ticks member).  To pause, I can
+disconnect the PWM peripheral, and  generate PWM interrupts instead of using
+the PIO interrupts.  It may seem like a good idea to just use the PWM
+interrupts.  However, this arrangement actually eliminates a branch, and keeps
+pause code seprate from pulse code.  On top of that, I get a true count of
+pulses output, rather than pulses attempted to be generated even if it was
+disconnected.  Its actually more accurate than my scope, as if I am in
+process of disconnecting the PWM peripheral, I still get a PIO interrupt, but
+on the scope, I see only a small .8-.9V pulse.  Its possible that pulse could
+be detected, though not likely.  Better to eliminate it if possible (which it
+was!)  The branch is eliminated by not needing to check if it was the cmpr
+register interrupt, or the period update interrupt.  The PWM interrupt is
+always the period update, and the PIO is always the cmp.  One branch isn't
+much, but its something!
+
+
+I think I'm going about this all wrong.  I think I can set the AB reg to the PWM
+generator, but set the output to 0 and still get interrupts on the PIO when a
+pulse would have happened if we hadn't overridden the output to 0.  Look into
+this...
+*/
+#include "FastAccelStepper.h"
+inline void disconnectPWMPeriphal(Pio* port, uint8_t pin, uint32_t channelMask);
+
+#define NUM_PWM_CHANNELS 8
+uint8_t TimerChannel_Map[NUM_PWM_CHANNELS];
+
+uint8_t numChannels = 0;
+
+// Here are the global variables to interface with the interrupts
+#if defined(KEEP_SCORE) //If you want to count how many pulses we detect...
+volatile uint32_t totalPulsesDetected[NUM_QUEUES];
+volatile uint32_t totalSteps[NUM_QUEUES];
+inline void IncrementQueue(int queue) { totalPulsesDetected[queue]++; }
+inline void AddToTotalSteps(int queue, uint32_t steps) { totalSteps[queue] += steps; }
+#else
+#define IncrementQueue(Q)
+#define AddToTotalSteps(Q, N)
+#endif
+bool channelsUsed[8] = {false, false, false, false, false, false, false, false};
+StepperQueue fas_queue[NUM_QUEUES];
+PWMCHANNELMAPPING gChannelMap[NUM_QUEUES];
+
+void TC5_Handler() {
+  uint32_t SR0 = TC1->TC_CHANNEL[2].TC_SR;
+  if (SR0 & TC_SR_CPCS) {
+    // We're going to use TC5 as its not connected to any pins on the DUE to run
+    // our engine... We only get in here on an RC compare, so if we're in here,
+    // run manage steppers...
+    fas_engine->manageSteppers();
+    TC1->TC_CHANNEL[2].TC_CCR = TC_CCR_SWTRG | TC_CCR_CLKEN;  // CLKEN and SWTRG
```

Comment:
In the avr code, the interrupts are enabled before calling `manageSteppers()`.
Reason is, that executing `manageSteppers()` in the interrupt context will block other interrupts on avr. How is this with due ?

Reply:
So the Arm Cortex M3 might not be the big beast that the STM32H series are, but it still has some impressive hardware.  The nested vectored interrupt controller is one such piece.  Its unlikely to be the same NVIC on the STM32H series, but it does already allow for interrupt nesting.  I intentionally set the timer 5 IRQ to a priority lower than the PIO IRQs.  However, that did bring me to where I never explicitly set the PWM IRQ.  (I believe 0 is default, but just like undefined behavior, suspected default isn't good enough and should be explicitly called out!)  I have had that adjustment in the code and commented some other code around it.  So manage steppers should be the lowest priority IRQ currently defined.  PIO/PWM are priority 0 (which is the highest priority), timer5 which runs manageSteppers is run at priority 1.  A user who needs this to be lower priority can easily call  NVIC_SetPriority(TC5_IRQn, lowerPriority);